### ISSUE TITLE
Bloop export issues with the plugin classpath

### ISSUE THREAD

**Issue by ckipp01** [com-lihaoyi/mill#1690]
So I'm not really sure how I haven't noticed this before, but it just bit me for the fist time and I'm trying to figure out where a change should happen. However I noticed today when looking at [`scala-cli`](https://scala-cli.virtuslab.org/) and trying to do a bloop export with the build that there were some errors being thrown in the logs about not being able to find the `scalac-plugin.xml` file. This was then seemingly causing issues with indexing and causing navigation not to work in the project.

_Example of the error_
```
2022.01.22 15:54:11 ERROR /Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.12/1.0.6/scala-xml_2.12-1.0.6.jar
java.nio.file.NoSuchFileException: /scalac-plugin.xml
```

I thought it was odd that Metals was looking for a `scalac-plugin.xml` for the `scala-xml` dep, but then I was able to figure out that the main issue was that the Bloop config that was being generated had everything in the `scalacPluginClasspath()` prefaced with `-Xplugin`. For example, here is what was generated for the cli module of `scala-cli` under the `scala` key:

```
"scala": {
    "organization": "org.scala-lang",
    "name": "scala-compiler",
    "version": "2.12.15",
    "options": [
        "-Xasync",
        "-Ywarn-unused",
        "-deprecation",
        "-Xplugin:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalamacros/paradise_2.12.15/2.1.1/paradise_2.12.15-2.1.1.jar",
        "-Xplugin:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalameta/semanticdb-scalac_2.12.15/4.4.31/semanticdb-scalac_2.12.15-4.4.31.jar",
        "-Xplugin:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-compiler/2.12.15/scala-compiler-2.12.15.jar",
        "-Xplugin:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.12.15/scala-library-2.12.15.jar",
        "-Xplugin:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.12.15/scala-reflect-2.12.15.jar",
        "-Xplugin:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.12/1.0.6/scala-xml_2.12-1.0.6.jar"
    ],
    "jars": [
        "/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-compiler/2.12.15/scala-compiler-2.12.15.jar",
        "/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.12.15/scala-reflect-2.12.15.jar",
        "/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.12.15/scala-library-2.12.15.jar",
        "/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.12/1.0.6/scala-xml_2.12-1.0.6.jar"
    ]
},
```

Compare this to another project I had up that also has similar deps under jars, but was an sbt project:

```
"scala": {
    "organization": "org.scala-lang",
    "name": "scala-compiler",
    "version": "2.12.15",
    "options": [
        "-target:jvm-1.8",
        "-Yrangepos",
        "-Ywarn-unused:imports",
        "-Ywarn-unused:privates",
        "-Ywarn-unused:locals",
        "-Ywarn-unused:patvars",
        "-Ywarn-unused:implicits",
        "-Xplugin:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalameta/semanticdb-scalac_2.12.15/4.4.32/semanticdb-scalac_2.12.15-4.4.32.jar"
    ],
    "jars": [
        "/Users/ckipp/.sbt/boot/scala-2.12.15/lib/scala-library.jar",
        "/Users/ckipp/.sbt/boot/scala-2.12.15/lib/scala-compiler.jar",
        "/Users/ckipp/.sbt/boot/scala-2.12.15/lib/scala-reflect.jar",
        "/Users/ckipp/.sbt/boot/scala-2.12.15/lib/scala-xml_2.12-1.0.6.jar"
    ],
```

Notice that only the actual plugin is listed with `-Xplugin`. From looking in the code it looks like it's happening in this task:

https://github.com/com-lihaoyi/mill/blob/5cb73cab162540963e2d50562dfed45512a6d60c/contrib/bloop/src/mill/contrib/bloop/BloopImpl.scala#L175-L179

Seeing that the output of the `scala-cli` is:

```
~/Documents/scala-workspace/scala-cli master
❯ ./mill show cli.scalacPluginClasspath
[1/1] show
[
    "qref:57d150fb:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalamacros/paradise_2.12.15/2.1.1/paradise_2.12.15-2.1.1.jar",
    "qref:81a5ff5a:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalameta/semanticdb-scalac_2.12.15/4.4.31/semanticdb-scalac_2.12.15-4.4.31.jar",
    "qref:e60597bd:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-compiler/2.12.15/scala-compiler-2.12.15.jar",
    "qref:9646cade:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.12.15/scala-library-2.12.15.jar",
    "qref:ba2f2029:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.12.15/scala-reflect-2.12.15.jar",
    "qref:22e89b01:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.12/1.0.6/scala-xml_2.12-1.0.6.jar"
]
```
It makes sense why this is happening, but I'm not 100% sure how or where this should be handled. When I resolve a plugin, like semanticdb for example:

```
❯ cs resolve org.scalameta:semanticdb-scalac_2.12.15:4.4.31
org.scala-lang:scala-library:2.12.15:default
org.scalameta:semanticdb-scalac_2.12.15:4.4.31:default
```
I see that the scala-library is there, but that really shouldn't be prefaced with `-Xplugin`, but I also haven't dug deep enough down yet to see if there is something funky with going on with how Mill is putting this together and why it's different than sbt, or if we are making too many assumptions in https://github.com/scalameta/metals/blob/main/metals/src/main/scala/scala/meta/internal/metals/CompilerPlugins.scala, but we haven't touched that code in quite some time.

I'm curious if you have any thoughts on this .

**Comment by ckipp01** [comment]
Ahhh, I realize now that we've hit on this before!

https://github.com/com-lihaoyi/mill/issues/880
https://github.com/scalameta/metals/issues/1720

I do remember this now thinking about it. 🤔 I feel like we should think about this a bit more and try to solve it since the workaround of:

```
.exclude("*" -> "*")
```
isn't the most user friendly in the world, and as I just experienced I even forgot about it 😆

**Comment by lefou** [comment]
Maybe, we are not supposed to resolve compiler plugins transitively? I could not found any information about this in the scala documentation. (We, @ckipp01 and me, also spoke about this before in context of scoverage plugin.)  I think most plugins do not require additional dependenciesm but thats just a guess. The best transitive dependency mapping for compiler plugins would be to declare all API dependencies, which should be provided by the compiler itself (like scala library jar and compiler jar) as Maven provided scope, or in Mill speak as compile-only dependencies. Unfortunately we don't have control over this so I see two options to improve the situation:

1. Never resolve compiler plugins transitively. This is essentially the same as providing each with `.exclude("*" -> "*")`. This will change behavior, as the resulting compiler classpath is likely smaller. This change might break compilations.

2. Only generate `-Xplugin` options from the directly listed ivy deps but still provide all transitive dependencies as compiler classpath. This will only generate less options, which seems the right thing but still provides all declared transitive dependencies. This will not break currently succeeding compilations, but might still violage the expectation of the compiler, as we potentially provide to much (conflicting) dependencies.

**Comment by lefou** [comment]
Thanks for reporting and providing such a in-depth analysis, btw.

**Comment by lolgab** [comment]
When there are this kind of problems, I always look what Sbt does, since it usually is the "source of truth" for Scala toolchains and considered "right" from people in the Scala community.
In the zinc internal code there is a [method to find the compiler plugins in a classpath](https://github.com/sbt/zinc/blob/425660612a2cdfb77fc8fcbb4463070a4d6c19d2/internal/zinc-classpath/src/main/scala/sbt/internal/inc/classpath/ClasspathUtilities.scala#L149).
It seems that what they do in Sbt is to [partition the compiler plugin classpath using that method](https://github.com/sbt/sbt/blob/a448b1caab5b028923c1390ab6cf23a1824d95dc/main/src/main/scala/sbt/Defaults.scala#L4096) and add the `"-Xplugin:"` prefix only to these ones.
We should probably do the same.

**Comment by ckipp01** [comment]
Hahaha, I was literally just looking at this @lolgab and no less than 10 minutes ago did I find the same `compilerPlugins` method which I was about to link to. I think this is exactly right, it makes sense to do this I think because in some scenarios eve when using "provided" we'll still hit on this. For example with the `scala-library`, it will sort of just be there.

**Comment by lefou** [comment]
Thank you both, Lorenzo and Chris, for your valuable insights. It looks like there is some magic involved. TBH, I don't like it personally. The reason is, that a compiler should also be usable without a build tool, and the invocation should be clear. That personal preference aside I try to motivate it.

I think we should only generate the `-Xplugin` option from the exactly given `scalacPluginIvyDeps`. This gives control to the user and is IMHO what the user wants. In the rare case, the user really wants to enable more that one compiler plugin with a single declared ivy dependency, she can just add the other ivy dependency too. But in case she does not want to enable an additional plugin and for some reasons a second plugin appears in the transitive dependency chain, she would be without options otherwise. Probably even given up before she realizes what her problem is.

To me, the only questions is, whether we should provide all the transitive dependencies to the compiler classpath or not. I guess the answer is yes, from looking at sbt code and the fact that we already do it.

TLTR, just generate a single `-Xplugin` option for each given dependency.

**Comment by lefou** [comment]
If you agree with that direction, I can prepare a PR.

**Comment by ckipp01** [comment]
I sort of just think that compiler plugins are a bit funky because there also seems to be different approaches and patterns out there.

> The reason is, that a compiler should also be usable without a build tool, and the invocation should be clear.

In theory it sort of is though right? Like there is an assumption that you can do `scalac -Xplugin:/path/to/jar file.scala` and it'll work, no build tool necessary. _However_ I also see the flipside where this doesn't always work out. Because like we are experiencing now:

1. Do you only included the actual dep as the `-Xplugin`?
2. Do you include the dep and all the transitve deps all as `-Xplugin`?
3. Or do you maybe concatenate the classpath and pass that into `-Xplugin`?

3 seems weird, but that's actually what I do do in sbt-scoverage since in that scenario I have a compiler plugin for Scala 2 that actually relies on a domain module. It's separated out because I also use it for the reporting module which can be used by Scala 3.

https://github.com/scoverage/sbt-scoverage/blob/73f8f92e41eddc2dad3a7404284b5aa47afaee52/src/main/scala/scoverage/ScoverageSbtPlugin.scala#L122-L141

I don't love that at all, and I'm maybe even breaking expectation by doing that, but I actually need to collect multiple jars and pass them in with `-Xplugin/first/path/to/jar:second/path/to/jar`.

So I think plugins are just tricky because it's hard to tell if 1, 2, or 3 is correct up above and many times the answer is probably you should just default to 1, and if there is a plugin that requires something funky like 3, then it's up to an integration or extension to correctly set it up.

**Comment by lefou** [comment]
Interesting details. To control the compiler plugins, we have to two entry points, the compiler classpath and the options. Where the former provides all the necessary jars, the latter enables a plugin which is probably already provided in the classpath. So I think my solution outlined above should just work for scoverage: all domain classes contained in the transitive dependencies are added to the compiler classpath as we fully transitively resolve the ivy dependency of the plugin, whereas the actual scoverage plugin is the only jar that gets added via the `-Xplugin:/path/to/scoverage/compiler/plugin.jar`.

In current implementation, we would end up with the same plus an `-Xplugin:` option for each of the transitive jars.

**Comment by lefou** [comment]
@ckipp01 From the sbt discussion you linked (https://github.com/sbt/sbt/issues/2255) I infer, that sbt is currently not resolving the transitive dependencies of a compiler plugin. Is that right?

I think at this point, I'd like to just implement my proposal in a proof of concept PR. Then we can add some test cases for known to be complicated plugins like scoverage.

**Comment by ckipp01** [comment]
> Interesting details. To control the compiler plugins, we have to two entry points, the compiler classpath and the options. Where the former provides all the necessary jars, the latter enables a plugin which is probably already provided in the classpath. So I think my solution outlined above should just work for scoverage: all domain classes contained in the transitive dependencies are added to the compiler classpath as we fully transitively resolve the ivy dependency of the plugin, whereas the actual scoverage plugin is the only jar that gets added via the `-Xplugin:/path/to/scoverage/compiler/plugin.jar`.

I'll be honest, I'm hitting on the edge of my knowledge on how this works, but this was something that actually confused me when I split apart scoverage that I still have yet to understand for figure out. I'm not 100% sure the above is actually correct. For example in sbt if I remove those extra jars from the actual `-Xplugin`, even though they are on the classpath, it explodes:

For example I just tested this and only include the actual compiler plugin jar like you said:
```
[info] a / Compile / compile / scalacOptions
[info]  List(-Xplugin:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scoverage/scalac-scoverage-plugin_2.12.15/2.0.0-M5/scalac-scoverage-plugin_2.12.15-2.0.0-M5.jar, -P:scoverage:dataDir:/Users/ckipp/Documents/scala-workspace/scoverage-test/a/target/scala-2.12/scoverage-data, -P:scoverage:sourceRoot:/Users/ckipp/Documents/scala-workspace/scoverage-test, -P:scoverage:reportTestName, -Yrangepos)
```
And the other jars for domain, serializer, and reporter are still included on the classpath it explodes:
```
[info] a / Compile / fullClasspath
[info]  List(Attributed(/Users/ckipp/Documents/scala-workspace/scoverage-test/a/target/scala-2.12/classes), Attributed(/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.12.15/scala-library-2.12.15.jar), Attributed(/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scoverage/scalac-scoverage-domain_2.12/2.0.0-M5/scalac-scoverage-domain_2.12-2.0.0-M5.jar), Attributed(/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scoverage/scalac-scoverage-reporter_2.12/2.0.0-M5/scalac-scoverage-reporter_2.12-2.0.0-M5.jar), Attributed(/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scoverage/scalac-scoverage-serializer_2.12/2.0.0-M5/scalac-scoverage-serializer_2.12-2.0.0-M5.jar), Attributed(/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scoverage/scalac-scoverage-runtime_2.12/2.0.0-M5/scalac-scoverage-runtime_2.12-2.0.0-M5.jar), Attributed(/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.12/1.0.6/scala-xml_2.12-1.0.6.jar))
```
```
[error] ## Exception when compiling 2 sources to /Users/ckipp/Documents/scala-workspace/scoverage-test/a/target/scala-2.12/classes
[error] java.lang.reflect.InvocationTargetException
[error] java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[error] java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[error] java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[error] java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
[error] scala.tools.nsc.plugins.Plugin$.instantiate(Plugin.scala:185)
[error] scala.tools.nsc.plugins.Plugins.$anonfun$loadRoughPluginsList$10(Plugins.scala:60)
....
```
Honestly the stacktrace is awful, but you can see that it has issues instantiating the plugin. This is what led me to actually concatenating it and putting it all behind `-Xplugin`.

**Comment by lefou** [comment]
I'm not an sbt expert neither, but `fullClasspath` is probably not the compiler classpath but the classpath needed to compile the actual source code.

**Comment by ckipp01** [comment]
> @ckipp01 From the sbt discussion you linked ([sbt/sbt#2255](https://github.com/sbt/sbt/issues/2255)) I infer, that sbt is currently not resolving the transitive dependencies of a compiler plugin. Is that right?

I'm assuming they are probably resolved, but not included. Test this I just threw this in a build file:
```scala
addCompilerPlugin("org.scalamacros" % "paradise" % "2.1.1" cross CrossVersion.patch)
```

Looking at the deps of this:
```
❯ cs resolve org.scalamacros:paradise_2.12.15:2.1.1
org.scala-lang:scala-compiler:2.12.15:default
org.scala-lang:scala-library:2.12.15:default
org.scala-lang:scala-reflect:2.12.15:default
org.scala-lang.modules:scala-xml_2.12:1.0.6:default
org.scalamacros:paradise_2.12.15:2.1.1:default
```
That seems to be stripped out when you look at the scalacOptions
```
sbt:hello-world> show scalacOptions
[info] * -Xplugin:/Users/ckipp/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scalamacros/paradise_2.12.15/2.1.1/paradise_2.12.15-2.1.1.jar
```

> I'm not an sbt expert neither, but fullClasspath is probably not the compiler classpath but the classpath needed to compile the actual source code.

Ahhhh you're correct. Dang, maybe that was my issue when `sbt-scoverage`. I can see this if I enable `coverage` and look at the `jars` section under the bloop json files for `scala`. They aren't there. So maybe I don't need to do that and like you said, just making sure they are passed to the compiler classpath should be enough. I should fix this in sbt-scoverage probably.

> I think at this point, I'd like to just implement my proposal in a proof of concept PR. Then we can add some test cases for known to be complicated plugins like scoverage.

Sounds good 👍🏼

**Comment by ckipp01** [comment]
Also just to bring this full circle around (sorry to talk about sbt so much in here), but it looks like the actual bloop exported jars under the `scala` key in sbt land are knows as the `ScalaInstance`. You can see how this comes together here in sbt:

https://github.com/sbt/sbt/blob/952546f413590f3b93a20b834b78d1e51e854ec7/main/src/main/scala/sbt/Defaults.scala#L1089-L1116

And then for Bloop when this is exported the jars are just grabbed from there:

https://github.com/scalacenter/bloop/blob/50cc7dc0efa736b0136673ebaac1c53fbbb72779/integrations/sbt-bloop/src/main/scala/bloop/integrations/sbt/SbtBloop.scala#L948-L949

I have a gut feeling that sbt doesn't want you touching the `ScalaInstance`. I did a little search to see if this is often added to, and it seems like it really isn't from what I can see:

https://sourcegraph.com/search?q=context:global+scalaInstance+lang:scala&patternType=literal

So this seems easy enough to do with mill by just adding to the `scalaCompilerClasspath`, but is this sort of a no no? I wonder if it's preferred by the compiler to do this, or to just concatenate them all with `-Xplugin`.

Maybe @SethTisue would have some insight into this?

**Comment by lefou** [comment]
@ckipp01 I opened PR #1705. CI is green, so it would be nice, if you could give this a try and report back.

**Comment by SethTisue** [comment]
> Maybe @SethTisue would have some insight into this?

Not really, beyond what's already on (or linked from) sbt/sbt#2255. I remember wanting to add something about it to https://docs.scala-lang.org/overviews/plugins/index.html but not knowing what exactly to add. The one time I faced this in a plugin of my own I dodged the question by making it a fat JAR (as documented at https://github.com/sbt/sbt-assembly/issues/178).
